### PR TITLE
LPS-88710 SchedulerEventMessageListener and SchedulerEngineHelper

### PR DESCRIPTION
**Pull request by dantewang**

### CODE REVIEW

**Comment by dantewang** [comment]
ci:test

**Comment by dantewang** [comment]
ci:retest

File: modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java [hunk 1]
```diff
@@ -997,9 +997,28 @@ public SchedulerEventMessageListener addingService(
...
-					properties);
+				if (schedulerEventMessageListener instanceof
+						SchedulerEventMessageListenerWrapper) {
+
+					serviceRegistration = bundleContext.registerService(
+						MessageListener.class, schedulerEventMessageListener,
+						properties);
+				}
+				else {
+					SchedulerEventMessageListenerWrapper
```
Comment (tinatian) [inline modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java:1008]: @dantewang @yuhai 
The logic in else here is to replace the logic in SchedulerEngineHelper.register, right?
But the logic here is not consistent with the logic in SchedulerEngineHelper.register. 

In register(), it is checking if current registration exists, but you did the registration without any checking. Is it safe to remove the checking? Did you test the logic here?


Thanks.
Tina.

**Comment by tinatian** [comment]
@dantewang @yuhai 
Please make sure you test scheduler manually to make sure you did not break anything, At least, you need to test redeploy:
1) Redeploy a module with scheduler, make sure old scheduler is stopped and new one is running
2) Redeploy a old war bundle with scheduler, make sure old scheduler is stopped and new one is running

Also, I would like to make the service tracker in SchedulerEngineHelperImpl more clean, could you please see what else you can do? The logic here is too complex and hard to maintain.

Thanks.
Tina.

File: modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java (hunk 1 above)
Comment (dantewang) [reply modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java:1008]: @tinatian The registration protection is already in https://github.com/tinatian/liferay-portal/blob/a52e7819b1c16afd5506a8618af0a96f5325fc21/modules/apps/portal-scheduler/portal-scheduler/src/main/java/com/liferay/portal/scheduler/internal/SchedulerEngineHelperImpl.java#L977

**Comment by dantewang** [comment]
@tinatian Hai did tests by modifying configs of the listeners (OSGi). Will do the manual test for deploy/redeploy.

For the service tracker, the incoming services can actually be divided into 2 types: `SchedulerEventMessageListener` and `SchedulerEventMessageListenerWrapper`, although they are all registered as `SchedulerEventMessageListener`.

The services of pure `SchedulerEventMessageListener` are the listeners we modified in this PR. They are mostly DS components.

The services that are actually wrapper instances are from `register` method (which third party code may still use) and https://github.com/tinatian/liferay-portal/pull/1147/commits/eb9847bfc3bcd22aec9f20fddd4f48b2cb0ac2d0.

So besides what is needed for scheduler, the complexity of the service tracker comes from:

1. To distinguish if the incoming service is pure listener or a wrapper; if it's a wrapper, register it directly; if not, wrap it and register the wrapped;
2. To update `SchedulerEntry` in the wrapper in `addingService` or `modifiedService`; since the wrapper wraps `MessageListener`, its `getSchedulerEntry` method is not a direct delegate.

**Comment by tinatian** [comment]
Hi @dantewang ,

Yes, like what you said, the problem is we support two ways to enable a scheduled job, one is to register a wrapper directly, another way is to register a listener.

Two ways to make it simpler:
1. Use two service trackers. One is to track pure listener and change is to a wrapper, another one is to track wrapper.
2. Make everyone register a pure listener directly instead of registering a wrapper, and register the wrapper internally later. In fact, the wrapper should be internal and only be known by scheduler itself, people out side of scheduler should not use it directly.

Tina.

**Pull request closed without merging**